encoding/codecv7.go

```diff
@@ -0,0 +1,446 @@
+	}
+
+	numL2Transactions := block.NumL2Transactions()
+	numTransactions := uint64(numL1Messages) + numL2Transactions
+	if numTransactions > math.MaxUint16 {
+		return nil, errors.New("number of transactions exceeds max uint16")
+	}
+
+	daBlock := newDABlockV7(
+		block.Header.Number.Uint64(), // number
+		block.Header.Time,            // timestamp
+		block.Header.BaseFee,         // baseFee
+		block.Header.GasLimit,        // gasLimit
+		uint16(numTransactions),      // numTransactions
+		numL1Messages,                // numL1Messages
+	)
+
+	return daBlock, nil
+}
+
+// NewDAChunk creates a new DAChunk from the given Chunk and the total number of L1 messages popped before.
+// Note: In DACodecV7 there is no notion of chunks. Blobs contain the entire batch data without any information of Chunks within.
+// However, for compatibility reasons this function is implemented to create a DAChunk from a Chunk.
+// This way we can still uniquely identify a set of blocks and their L1 messages.
+func (d *DACodecV7) NewDAChunk(chunk *Chunk, totalL1MessagePoppedBefore uint64) (DAChunk, error) {
+	if chunk == nil {
+		return nil, errors.New("chunk is nil")
+	}
+
+	if len(chunk.Blocks) == 0 {
+		return nil, errors.New("number of blocks is 0")
+	}
+
+	if len(chunk.Blocks) > math.MaxUint8 {
+		return nil, fmt.Errorf("number of blocks (%d) exceeds maximum allowed (%d)", len(chunk.Blocks), math.MaxUint8)
+	}
+
+	initialL2BlockNumber := chunk.Blocks[0].Header.Number.Uint64()
+	l1MessageIndex := totalL1MessagePoppedBefore
+
+	blocks := make([]DABlock, 0, len(chunk.Blocks))
+	txs := make([][]*types.TransactionData, 0, len(chunk.Blocks))
+
+	for i, block := range chunk.Blocks {
+		// sanity check: block numbers are contiguous
+		if block.Header.Number.Uint64() != initialL2BlockNumber+uint64(i) {
+			return nil, fmt.Errorf("invalid block number: expected %d but got %d", initialL2BlockNumber+uint64(i), block.Header.Number.Uint64())
+		}
+
+		// sanity check (within NumL1MessagesNoSkipping): L1 message indices are contiguous within a block
+		numL1Messages, highestQueueIndex, err := block.NumL1MessagesNoSkipping()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get numL1Messages: %w", err)
+		}
+		// sanity check: L1 message indices are contiguous across blocks boundaries
+		if numL1Messages > 0 {
+			if l1MessageIndex+uint64(numL1Messages) != highestQueueIndex+1 {
+				return nil, fmt.Errorf("failed to sanity check L1 messages count after block %d: l1MessageIndex + numL1Messages != highestQueueIndex+1: %d + %d != %d", block.Header.Number.Uint64(), l1MessageIndex, numL1Messages, highestQueueIndex+1)
+			}
+			l1MessageIndex += uint64(numL1Messages)
+		}
+
+		daBlock := newDABlockV7(block.Header.Number.Uint64(), block.Header.Time, block.Header.BaseFee, block.Header.GasLimit, uint16(len(block.Transactions)), numL1Messages)
```

Comment:
Why not reuse `NewDABlock` here? Is it because you want to avoid calling `NumL1MessagesNoSkipping` twice?

Reply:
Yes I wanted to avoid that. But I think we don't actually need to call it again. Changed in de09af4bc445412cf6fb4933a0213143b54b8b23 to use `NewDABlock`